### PR TITLE
chore: add new test (already passing) around function types & no-restricted-globals rule

### PR DESCRIPTION
**Pull request by Andarist**
I'm experiencing this rule to fail on similar code and I was about to raise an issue - I've tested first if it works here and it seems it does, so my issue has to be connected to my setup and need to fix it on my own.

Nevertheless, I couldn't verify quickly if that simple case is handled with the plugin here (had to check "manually" by cloning the repo and adding the test case myself) - so I thought I might send this as a PR, maybe it can help a future somebody.

If you feel this excessive then feel free to just close this.

I would have a question though, regarding this - how this plugin works exactly for rules like this? Does it hook into default ESLint scope analysis? Does it override the default behavior somehow?

### CODE REVIEW

**Comment by bradzacher** [comment]
> Does it hook into default ESLint scope analysis?

We don't hook into it, we define our own.
The test you modified is a sanity check to ensure we don't break the rule with those changes.

**Comment by Andarist** [comment]
@bradzacher  Could u point me to the code lines where this extending takes place? I would be interested in taking a look how u handle that.

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/parser/src/analyze-scope.ts

tl;dr - we do it pretty poorly right now.
There was a maintainer that was working on it at the start of the year, but they stepped back from the project not long after. Nobody has had the time to pick it up since then.
There is a lot that you have to figure out about how scope analysis works in eslint, as the package that's used (https://github.com/eslint/eslint-scope) is mostly undocumented.
That's the main reason I haven't gone back to finish the work.

As an aside - this is why rules like `no-unused-vars`, `no-use-before-define`, etc that rely on correct scope information are broken - because the typescript scope analysis code isn't complete.
